File: src/lib/api/agents.ts
/**
 * Agent registry IPC wrappers. Browser mocks return empty/defaults when not in Tauri.
 */
import { invoke } from "@tauri-apps/api/core";

function isTauri(): boolean {
  return typeof window !== "undefined" && "__TAURI_INTERNALS__" in window;
}

export type Agent = {
  id: string;
  name: string;
  description: string | null;
  workspace_path: string;
  git_url: string | null;
  default_cli: string;
  status: string;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
};

export type AgentUpsert = {
  id?: string | null;
  name: string;
  description?: string | null;
  workspace_path: string;
  git_url?: string | null;
  default_cli: string;
  status?: string | null;
};

export type AgentModelProfile = {
  agent_id: string;
  preferred_model: string | null;
  reasoning_effort: string | null;
  engine_options_json: string | null;
};

export type Skill = {
  id: string;
  agent_id: string;
  relative_path: string;
  title: string | null;
  description: string | null;
  enabled: boolean;
  content_hash: string | null;
  scanned_at: string | null;
};

export type SkillUpsert = {
  id?: string | null;
  agent_id: string;
  relative_path: string;
  title?: string | null;
  description?: string | null;
  enabled?: boolean | null;
  content_hash?: string | null;
};

export type ImportAgentResult = {
  agent: Agent;
  cloned: boolean;
  workspace_path: string;
  skills_synced?: {
    added: number;
    updated: number;
    removed: number;
  } | null;
  skills_sync_error?: string | null;
};

export async function listAgents(): Promise<Agent[]> {
  if (!isTauri()) return [];
  return invoke<Agent[]>("list_agents");
}

export async function importAgent(input: {
  name: string;
  workspace_path_or_git: string;
  default_cli: string;
  description?: string | null;
}): Promise<ImportAgentResult> {
  if (!isTauri()) {
    throw new Error("导入 Agent 需要桌面应用（Tauri）运行时");
  }
  return invoke<ImportAgentResult>("import_agent", {
    name: input.name,
    workspacePathOrGit: input.workspace_path_or_git,
    defaultCli: input.default_cli,
    description: input.description ?? null,
  });
}

export async function upsertAgent(agent: AgentUpsert): Promise<Agent> {
  if (!isTauri()) {
    throw new Error("保存 Agent 需要桌面应用（Tauri）运行时");
  }
  return invoke<Agent>("upsert_agent", { agent });
}

export async function deleteAgent(id: string): Promise<void> {
  if (!isTauri()) return;
  await invoke("delete_agent", { id });
}

export async function getAgentProfile(
  agentId: string,
): Promise<AgentModelProfile | null> {
  if (!isTauri()) return null;
  return invoke<AgentModelProfile | null>("get_agent_profile", {
    agentId,
  });
}

export async function upsertAgentProfile(
  profile: AgentModelProfile,
): Promise<AgentModelProfile> {
  if (!isTauri()) return profile;
  return invoke<AgentModelProfile>("upsert_agent_profile", { profile });
}

export async function listSkills(agentId: string): Promise<Skill[]> {
  if (!isTauri()) return [];
  return invoke<Skill[]>("list_skills", { agentId });
}

export async function upsertSkills(skills: SkillUpsert[]): Promise<Skill[]> {
  if (!isTauri()) return [];
  return invoke<Skill[]>("upsert_skills", { skills });
}

export async function setSkillEnabled(
  id: string,
  enabled: boolean,
): Promise<Skill> {
  if (!isTauri()) {
    throw new Error("切换 Skill 需要桌面应用（Tauri）运行时");
  }
  return invoke<Skill>("set_skill_enabled", { id, enabled });
}

export type SyncSkillsResult = {
  added: number;
  updated: number;
  removed: number;
};

export async function syncAgentSkills(
  agentId: string,
): Promise<SyncSkillsResult> {
  if (!isTauri()) return { added: 0, updated: 0, removed: 0 };
  return invoke<SyncSkillsResult>("sync_agent_skills", { agentId });
}

export async function readSkillContent(
  agentId: string,
  relativePath: string,
): Promise<string> {
  if (!isTauri()) {
    throw new Error("读取 Skill 需要桌面应用（Tauri）运行时");
  }
  return invoke<string>("read_skill_content", {
    agentId,
    relativePath,
  });
}

File: src-tauri/src/commands/agents.rs
use crate::repo::{
    delete_agent as repo_delete_agent, get_agent_profile as repo_get_profile,
    list_agents as repo_list_agents, list_skills_by_agent, set_skill_enabled as repo_set_skill,
    upsert_agent as repo_upsert_agent, upsert_agent_profile as repo_upsert_profile,
    upsert_skills_many, Agent, AgentModelProfile, AgentUpsert, Skill, SkillUpsert,
};
use crate::services::import_agent::{
    import_agent as service_import_agent, ImportAgentRequest, ImportAgentResult,
};
use crate::services::skill_scan::{
    read_skill_content as service_read_skill, sync_agent_skills as service_sync_skills,
    SyncSkillsResult,
};
use crate::state::DbState;
use tauri::State;

fn with_db<T, F>(state: &State<'_, DbState>, f: F) -> Result<T, String>
where
    F: FnOnce(&rusqlite::Connection) -> Result<T, String>,
{
    let conn = state
        .conn
        .lock()
        .map_err(|e| format!("db lock poisoned: {e}"))?;
    f(&conn)
}

#[tauri::command]
pub fn list_agents(state: State<'_, DbState>) -> Result<Vec<Agent>, String> {
    with_db(&state, repo_list_agents)
}

#[tauri::command]
pub fn import_agent(
    state: State<'_, DbState>,
    name: String,
    workspace_path_or_git: String,
    default_cli: String,
    description: Option<String>,
) -> Result<ImportAgentResult, String> {
    with_db(&state, |c| {
        service_import_agent(
            c,
            ImportAgentRequest {
                name,
                workspace_path_or_git,
                default_cli,
                description,
            },
        )
    })
}

#[tauri::command]
pub fn upsert_agent(state: State<'_, DbState>, agent: AgentUpsert) -> Result<Agent, String> {
    with_db(&state, |c| repo_upsert_agent(c, agent))
}

#[tauri::command]
pub fn delete_agent(state: State<'_, DbState>, id: String) -> Result<(), String> {
    with_db(&state, |c| repo_delete_agent(c, &id))
}

#[tauri::command]
pub fn get_agent_profile(
    state: State<'_, DbState>,
    agent_id: String,
) -> Result<Option<AgentModelProfile>, String> {
    with_db(&state, |c| repo_get_profile(c, &agent_id))
}

#[tauri::command]
pub fn upsert_agent_profile(
    state: State<'_, DbState>,
    profile: AgentModelProfile,
) -> Result<AgentModelProfile, String> {
    with_db(&state, |c| repo_upsert_profile(c, profile))
}

#[tauri::command]
pub fn list_skills(state: State<'_, DbState>, agent_id: String) -> Result<Vec<Skill>, String> {
    with_db(&state, |c| list_skills_by_agent(c, &agent_id))
}

#[tauri::command]
pub fn upsert_skills(
    state: State<'_, DbState>,
    skills: Vec<SkillUpsert>,
) -> Result<Vec<Skill>, String> {
    with_db(&state, |c| upsert_skills_many(c, &skills))
}

#[tauri::command]
pub fn set_skill_enabled(
    state: State<'_, DbState>,
    id: String,
    enabled: bool,
) -> Result<Skill, String> {
    with_db(&state, |c| repo_set_skill(c, &id, enabled))
}

#[tauri::command]
pub fn sync_agent_skills(
    state: State<'_, DbState>,
    agent_id: String,
) -> Result<SyncSkillsResult, String> {
    with_db(&state, |c| service_sync_skills(c, &agent_id))
}

#[tauri::command]
pub fn read_skill_content(
    state: State<'_, DbState>,
    agent_id: String,
    relative_path: String,
) -> Result<String, String> {
    with_db(&state, |c| service_read_skill(c, &agent_id, &relative_path))
}
